### PR TITLE
fix(2058): shared command expiry

### PR DESCRIPTION
**Pull request by mchengat**
## Context

Shared commands expire due to a side effect of using catbox-s3

## Objective

Avoid using catbox-s3 for commands

## References

https://github.com/screwdriver-cd/screwdriver/issues/2058

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: helpers/aws.js [hunk 1]
```diff
@@ -157,6 +157,35 @@ class AwsClient {
...
 
+    /**
+     * Upload command as Stream
+     * @method uploadCmdAsStream
+     * @param {Object}             config                Config object
+     * @param {Buffer}              config.payload       Payload to upload
+     * @param {String}              config.objectKey       Path to Object
+     * @return {Promise}
+     */
+    uploadCmdAsStream({ payload, objectKey }) {
```
Comment (jithine) [inline helpers/aws.js:168]: maybe just call it `uploadCmdAsStream` it's generic enough right ?

File: plugins/commands.js [hunk 2]
```diff
@@ -139,15 +177,32 @@ exports.plugin = {
                 const { namespace, name, version } = request.params;
                 const id = `${namespace}-${name}-${version}`;
 
-                try {
-                    await cache.drop(id);
-                    request.log([id, 'info'], 'Successfully deleted a command');
+                if (usingS3) {
+                    try {
```
Comment (jithine) [inline plugins/commands.js:181]: moving the try .. catch out of the if block can reduce the repeated lines

File: plugins/commands.js [hunk 3]
```diff
@@ -34,22 +45,37 @@ exports.plugin = {
...
+                    if (!value) {
+                        throw boom.notFound();
+                    }
 
-                response.headers = value.h;
+                    if (value.c) {
+                        response = h.response(Buffer.from(value.c.data));
+                        response.headers = value.h;
+                    } else {
+                        response = h.response(Buffer.from(value));
```
Comment (jithine) [inline plugins/commands.js:75]: where is this change coming from ?

File: plugins/commands.js [hunk 4]
```diff
@@ -34,22 +45,37 @@ exports.plugin = {
...
-                }
-
-                if (!value) {
-                    throw boom.notFound();
-                }
+                if (usingS3) {
+                    try {
+                        value = await awsClient.getDownloadStream({ cacheKey: id });
+                        response = h.response(value);
+                        response.headers['content-type'] = 'application/octet-stream';
```
Comment (jithine) [inline plugins/commands.js:55]: is this the current content-type returned when used with catbox-s3 ?

File: helpers/aws.js [hunk 5]
```diff
@@ -157,6 +157,35 @@ class AwsClient {
...
+        // stream the data to s3
+        const passthrough = new stream.PassThrough();
```
Comment (tkyi) [inline helpers/aws.js:170]:
```suggestion
        const passThrough = new stream.PassThrough();
```

File: helpers/aws.js [hunk 6]
```diff
@@ -195,6 +224,19 @@ class AwsClient {
             return s3stream;
         });
     }
+
+    deleteObject(object, callback) {
```
Comment (tkyi) [inline helpers/aws.js:228]: Can you add jsdocs

File: plugins/commands.js [hunk 7]
```diff
@@ -21,6 +23,15 @@ exports.plugin = {
      * @param  {Integer}  options.maxByteSize   Maximum Bytes to accept
      */
     register(server, options) {
+        const segment = 'commands';
```
Comment (tkyi) [inline plugins/commands.js:26]: Maybe set to constant?

**Comment by tkyi** [comment]
No test change?

File: plugins/commands.js (hunk 3 above)
Comment (mchengat) [reply plugins/commands.js]: This condition is not required, removed it.

File: plugins/commands.js (hunk 4 above)
Comment (mchengat) [reply plugins/commands.js:55]: yes, return content type is application/octet-stream

File: helpers/aws.js (hunk 5 above)
Comment (mchengat) [reply helpers/aws.js]: fixed

File: helpers/aws.js (hunk 6 above)
Comment (mchengat) [reply helpers/aws.js:234]: Done

File: plugins/commands.js (hunk 7 above)
Comment (mchengat) [reply plugins/commands.js:26]: @tkyi do you mean to update const segment = 'commands'

Comment (tkyi) [reply plugins/commands.js:26]: sorry I meant like this: 
```suggestion
        const SEGMENT = 'commands';
```
maybe move as global var

**Review by tkyi — COMMENTED** [review]
Should this be removed? https://github.com/screwdriver-cd/store/blob/master/bin/server#L13

**Review by jithine — APPROVED** [review]

File: plugins/commands.js (hunk 7 above)
Comment (mchengat) [reply plugins/commands.js:26]: kept it same as https://github.com/screwdriver-cd/store/blob/44eb3492c83602ddf0bcad034482ab9b46b5f7a9/plugins/builds.js#L34

**Comment by mchengat** [comment]
https://github.com/screwdriver-cd/store/pull/111#pullrequestreview-624550294
as cache is used by build, caches etc we can keep it like that

**Pull request merged**